src/java.base/share/classes/javax/net/ssl/ExtendedSSLSession.java

```diff
@@ -163,4 +164,99 @@ public List<SNIServerName> getRequestedServerNames() {
+     *
+     * @throws SSLKeyException if the key could not be generated
+     * @throws IllegalArgumentException if {@code length} is non-positive,
+     *         or if the {@code label} or {@code context} length can
+     *         not be accommodated
+     * @throws NullPointerException if {@code label} is null
+     * @throws UnsupportedOperationException if the underlying provider
+     *         does not implement the operation
+     *
+     * @return a {@code SecretKey} that contains {@code length} bytes of the
+     *         EKM material.
+     *
+     * @since 25
+     */
+    public SecretKey exportKeyingMaterialKey(
+            String label, byte[] context, int length) throws SSLKeyException {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Generate Exported Keying Material (EKM) calculated according to the
+     * algorithms defined in RFCs 5705/8446.
+     * <P>
+     * Note RFC 5705 (for (D)TLSv1.2 and earlier) calculates different EKM
+     * values depending on whether {@code context} is null or non-null/empty.
+     * RFC 8446 (TLSv1.3) treats a null context as non-null/empty.
+     * <P>
+     * The {@code label} {@code String} will be converted to bytes using
+     * the {@link java.nio.charset.StandardCharsets#UTF_8}
+     * character encoding.
+     * <P>
+     * Depending on the chosen underlying key derivation mechanism, the
+     * raw bytes might not be extractable/exportable.  In such cases, the
+     * {@link #exportKeyingMaterialKey(String, byte[], int)} method should be
+     * used instead to access the generated key material.
+     *
+     * @spec https://www.rfc-editor.org/info/rfc5705
+     *     RFC 5705: Keying Material Exporters for Transport Layer
+     *     Security (TLS)
+     * @spec https://www.rfc-editor.org/info/rfc8446
+     *     RFC 8446: The Transport Layer Security (TLS) Protocol Version 1.3
+     *
+     * @implSpec The default implementation throws
+     *           {@code UnsupportedOperationException}
+     *
+     * @param label   the label bytes used in the EKM calculation.
+     *                {@code label} will be converted to a {@code byte[]}
+     *                before the operation begins
+     * @param context the context bytes used in the EKM calculation
+     * @param length  the number of bytes of EKM material needed
+     *
+     * @throws SSLKeyException if the key could not be generated
+     * @throws IllegalArgumentException if {@code length} is non-positive,
+     *         or if the {@code label} or {@code context} length can
+     *         not be accommodated
+     * @throws NullPointerException if {@code label} is null
+     * @throws UnsupportedOperationException if the underlying provider
+     *         does not implement the operation
+     *
+     * @return a byte array of size {@code length} that contains the EKM
+     *         material, or null if the derived key material does not support
```

Comment:
For "or null if the derived key material does not support encoding" why wouldn't an implementation throw UOE instead?

Reply:
I was following the SecretKey.getEncoded() style.  I see now that KDF.deriveData() does do UOE.  

I could go either way on this.  I realize I need to make this consistent, I have TLSv1.3 using KDF style, and TLSv1-TLSv1.2 using the null.

Reply:
It seems like it should be an exception, whatever you decide to do. The caller is asking for the keying material data, and the provider cannot fulfill that request, so I think explaining why it could not be done would be best represented in an exception.

Reply:
+1 for UOE. It's not that we got a key and found it un-extractable. It's that we asked for data but were refused.

Reply:
I have updated to return a `UOE`.  

I'm a little concerned that `UOE` will be returned for both a non-overriden `exportKeyingMaterial*` method, and if there is a problem with the derivation.